### ISSUE TITLE
Support private gem repositories for plugins

### ISSUE THREAD

**Issue by purbon** [elastic/logstash#3576]
The current implementation of the command ```bin/plugin [install]``` ignores any gem repo that is not rubygems.

To reproduce this I used https://github.com/geminabox/geminabox, build a plugin myself, that is not released at rubygems, add the new gem source in several ways, like

* gem sources -a [gem_server]
* Adding it below the line ```source 'https:///rubygems.org'``` in the Gemfile. and update the Gemfile.lock
* Make it the only possible source in the Gemfile, also updating the Gemfile.lock.

If we aim to support private gem repositories, we should load all sources. 

Related: #2376

**Comment by purbon** [comment]
Adding the new source manually around https://github.com/elastic/logstash/blob/master/lib/pluginmanager/main.rb#L12 lets the installation to finish.

**Comment by colinsurprenant** [comment]
@purbon not sure I understand your last comment about adding the new source in `main.rb` ?!

In any case, manually adding `source "{source_url}"` in the `Gemfile` *should* work... if it doesn't I think this is where we should focus for a fix.

**Comment by purbon** [comment]
@colinsurprenant my last comment was just a note that adding this, while debugging the problem got the plugin installed, not a proposed solution at all.

manually adding the source to the Gemfile does not work to install plugins that are *only* in the private repo. But I agree adding a new source is the way to fix this, we just need to find a proper way to tell bundler about it.

**Comment by purbon** [comment]
Clarification:

the process that fails is the verification, not the installation perse.

```
─purbon@purbon-elastic  ~/work/logstash  ‹fix/plugin_manager_and_private_repo*› 
╰─$ bin/plugin install logstash-input-log4r 
Validating logstash-input-log4r
Plugin logstash-input-log4r does not exist
ERROR: Installation aborted, verification failed for logstash-input-log4r 
╭─purbon@purbon-elastic  ~/work/logstash  ‹fix/plugin_manager_and_private_repo*› 
╰─$ bin/plugin install --no-verify logstash-input-log4r                                                                                   1 ↵
Installing logstash-input-log4r
Installation successful
```

**Comment by colinsurprenant** [comment]
@purbon AH!!! yes, makes sense... the verification directly talks to `rubygems.org` regardless of the sources... good catch.

**Comment by purbon** [comment]
@colinsurprenant made a proposal to fix this in #3583, feel free to review when you get a minute.

**Comment by purbon** [comment]
wondering if makes sense to enable a couple of notations in our Gemfile, something like:

``` gem foo, :source => bar ```

and

```
source foo do
     gem bar
   end
```

Both are valid bundler, rubygems notation. I guess the first one makes a lot of sense, so people can then fetch selectively custom versions of plugins. What do you think?

**Comment by colinsurprenant** [comment]
`gem foo, :source => bar` will be correctly parsed and should be good for the intallation phase but will result in the same verification problem, I guess in that case we'd have to add logic to handle verification source per plugin/gem.

For the block style syntax lile
```
source foo do
  gem bar
end
```

it is a bit more complicated to handle. Our Gemfile parser accepts a subset of the valid Rubygems notation and it would have to be refactored to accept this syntax.

**Comment by purbon** [comment]
yeah, makes sense. However, I don't think the first syntax would survey an update (https://github.com/elastic/logstash/blob/master/lib/pluginmanager/gemfile.rb#L46-L48)

**Comment by purbon** [comment]
Fixed at #3583
